R/create_single_dose_dataset.R

```diff
@@ -0,0 +1,384 @@
+#'
+#'
+#' @details Each aggregate dose row is split into multiple rows which each represent
+#' a single dose.The number of completed dose periods between `start_date` and
+#' `end_date` is calculated with `compute_duration` and multiplied by `DOSE_COUNT`.
+#' For `DOSE_WINDOW` values of `"WEEK"`, `"MONTH"`, and `"YEAR"`, `CONVERSION_FACTOR`
+#' is used to convert into days the time object to be added to `start_date`.
+#'
+#' @author Michael Thorpe, Andrew Smith
+#'
+#' @keywords adae adex user_utility
+#'
+#' @return The input dataset with a single dose per row.
+#'
+#' @export
+#'
+#' @examples
+#' # Example with default lookup
+#'
+#' library(lubridate)
+#'
+#' data <- tibble::tribble(
+#'   ~USUBJID, ~EXDOSFRQ, ~ASTDT, ~AENDT,
+#'   "P01", "Q2D", ymd("2021-01-01"), ymd("2021-01-07"),
+#'   "P01", "Q3D", ymd("2021-01-08"), ymd("2021-01-15"),
+#'   "P01", "EVERY 2 WEEKS", ymd("2021-01-15"), ymd("2021-01-29")
+#' )
+#'
+#' create_single_dose_dataset(data)
+#'
+#' # Example with custom lookup
+#'
+#' custom_lookup <- tibble::tribble(
+#'   ~Value,  ~DOSE_COUNT, ~DOSE_WINDOW, ~CONVERSION_FACTOR,
+#'   "Q30MIN", (1 / 30), "MINUTE", 1,
+#'   "Q90MIN", (1 / 90), "MINUTE", 1
+#' )
+#'
+#' data <- tibble::tribble(
+#'   ~USUBJID, ~EXDOSFRQ, ~ASTDTM, ~AENDTM,
+#'   "P01", "Q30MIN", ymd_hms("2021-01-01T06:00:00"), ymd_hms("2021-01-01T07:00:00"),
+#'   "P02", "Q90MIN", ymd_hms("2021-01-01T06:00:00"), ymd_hms("2021-01-01T09:00:00")
+#' )
+#'
+#' create_single_dose_dataset(data,
+#'                            lookup_table = custom_lookup,
+#'                            lookup_column = Value,
+#'                            start_date = ASTDTM,
+#'                            end_date = AENDTM)
+
+create_single_dose_dataset <- function(dataset,
+                                       dose_freq = EXDOSFRQ,
+                                       start_date = ASTDT,
+                                       end_date = AENDT,
+                                       lookup_table = dose_freq_lookup,
+                                       lookup_column = CDISC_VALUE) {
+
+  col_names <- colnames(dataset)
+  dose_freq <- assert_symbol(enquo(dose_freq))
+  lookup_column <- assert_symbol(enquo(lookup_column))
+  start_date <- assert_symbol(enquo(start_date))
+  end_date <- assert_symbol(enquo(end_date))
+  assert_data_frame(dataset, required_vars = quo_c(dose_freq, start_date, end_date))
+  assert_data_frame(lookup_table, required_vars = vars(DOSE_WINDOW, DOSE_COUNT, CONVERSION_FACTOR))
+
+  # Set up lookup table to be joined to dataset
+
+  lookup <- lookup_table %>%
+    rename(!!dose_freq := !!lookup_column)
+
+  # Check values of lookup vs. data and return error if values are not covered
+
+  value_check <- dataset %>%
+    select(!!dose_freq) %>%
+    anti_join(lookup, by = as.character(quo_get_expr(dose_freq))) %>%
+    unique()
+
+  if (nrow(value_check) > 0) {
+    values_not_found <- str_c(value_check %>% select(!!dose_freq), collapse = ", ")
```

Comment:
This causes a warning:
```
In addition: Warning message:
In stri_c(..., sep = sep, collapse = collapse, ignore_null = TRUE) :
  argument is not an atomic vector; coercing
```
E.g., when calling
```
tex <- tibble::tribble(
  ~USUBJID, ~ASTDTM, ~AENDTM, ~EXDOSFRQ, ~EXTRT, ~temp_obs_nr,
  "1", ymd_hms("2020-02-07T02:00:00"), ymd_hms("2020-02-09T01:00:00"), "QUID", "DRUG 1", 1,
  "1", ymd_hms("2020-02-07T02:00:00"), ymd_hms("2020-02-09T01:00:00"), "QUOD", "DRUG 2", 2
) %>%
  mutate(ASTDT = date(ASTDTM),
         AENDT = date(AENDTM))

create_single_dose_dataset(tex,
                           start_date = ASTDTM,
                           end_date = AENDTM)
```

Reply:
switched to `paste0` and seems to be fine now